mixer/adapter/fluentd/fluentd_test.go

```diff
@@ -0,0 +1,281 @@
+			b.SetAdapterConfig(&c.config)
+
+			ce := b.Validate()
+			if (ce != nil) && c.success {
+				t.Errorf("Got %v, expecting success", ce)
+			} else if (ce == nil) && !c.success {
+				t.Errorf("Got success, expecting failure")
+			}
+
+			var h adapter.Handler
+			var err error
+			if c.inject {
+				h, err = b.injectBuild(context.Background(), env, &fluent.Fluent{})
+			} else {
+				h, err = b.Build(context.Background(), env)
+			}
+			if (err != nil) && c.success {
+				t.Errorf("Got %v, expecting success", err)
+			} else if (err == nil) && !c.success {
+				t.Errorf("Got success, expecting failure")
+			}
+			if (h == nil) && c.success {
+				t.Errorf("Got nil, expecting valid handler")
+			} else if (h != nil) && !c.success {
+				t.Errorf("Got a handler, expecting nil")
+			}
+		})
+	}
+}
+
+func TestLogEntry(t *testing.T) {
+	types := map[string]*logentry.Type{
+		"Foo": {
+			Variables: map[string]descriptor.ValueType{
+				"String":    descriptor.STRING,
+				"Int64":     descriptor.INT64,
+				"Double":    descriptor.DOUBLE,
+				"Bool":      descriptor.BOOL,
+				"Duration":  descriptor.DURATION,
+				"Time":      descriptor.TIMESTAMP,
+				"StringMap": descriptor.STRING_MAP,
+				"IPAddress": descriptor.IP_ADDRESS,
+				"Bytes":     descriptor.VALUE_TYPE_UNSPECIFIED,
+			},
+		},
+	}
+
+	mf := &mockFluentd{}
+
+	han := &handler{
+		logger: mf,
+		types:  types,
+		env:    test.NewEnv(t),
+	}
+
+	tm := time.Date(2017, time.August, 21, 10, 4, 00, 0, time.UTC)
+
+	cases := []struct {
+		instances    []*logentry.Instance
+		failWrites   bool
+		expectedTags []string
+		expected     []map[string]interface{}
+	}{
+		{
+			[]*logentry.Instance{
+				{
+					Name:     "Foo",
+					Severity: "WARNING",
+					Variables: map[string]interface{}{
+						"String": "a string",
+					},
+				},
+				{
+					Name:     "Foo",
+					Severity: "WARNING",
+					Variables: map[string]interface{}{
+						"tag": "fluent-tag",
+					},
+				},
+			},
+			false,
+			[]string{
+				"Foo",
+				"fluent-tag",
+			},
+			[]map[string]interface{}{
+				{
+					"severity": "WARNING",
+					"String":   "a string",
+				},
+				{
+					"name":     "Foo",
+					"severity": "WARNING",
+				},
+			},
+		},
+
+		{
+			[]*logentry.Instance{
+				{
+					Name:      "Foo",
+					Severity:  "WARNING",
+					Timestamp: tm,
+					Variables: map[string]interface{}{
+						"String":    "a string",
+						"Int64":     int64(123),
+						"Double":    1.23,
+						"Bool":      true,
+						"Time":      tm,
+						"Duration":  1 * time.Second,
+						"StringMap": map[string]string{"A": "B", "C": "D"},
+						"IPAddress": net.IPv4zero,
+						"Bytes":     []byte{'b'},
+					},
+				},
+			},
+			false,
+			[]string{
+				"Foo",
+			},
+			[]map[string]interface{}{
+				{
+					"severity": "WARNING",
+					"String":   "a string",
+					"Duration": "1s",
+				},
+			},
+		},
+	}
+
+	for i, c := range cases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
```

Comment:
as above, let's give these things reasonable names. the cases are complex enough that a readable name for each case would help others debug a bit.

Reply:
Done